src/payment/coupon/data/sagas.test.js

```diff
@@ -0,0 +1,246 @@
+        {
+          post: apiClientPost,
+        },
+      );
+
+      const dispatched = [];
+      await runSaga(
+        { dispatch: action => dispatched.push(action) },
+        handleAddCoupon,
+        addCoupon('DEMO25'),
+      ).toPromise();
+
+      expect(dispatched).toEqual([
+        addCouponBegin(),
+        addCouponSuccess(12345, 'DEMO25', {
+          type: PERCENTAGE_BENEFIT,
+          value: 25,
+        }),
+      ]);
+    });
+
+    it('should handle an API error', async () => {
+      const errorResponse = {
+        response: {
+          data: {
+            error_code: 'uhoh',
+          },
+        },
+      };
+
+      const apiClientPost = jest.fn(() =>
+        new Promise((resolve, reject) => {
+          reject(errorResponse);
+        }));
+
+      configureApiService(
+        {
+          ECOMMERCE_BASE_URL: 'http://localhost',
+        },
+        {
+          post: apiClientPost,
+        },
+      );
+
+      const dispatched = [];
+      await runSaga(
+        {
+          dispatch: action => dispatched.push(action),
+        },
+        handleAddCoupon,
+        addCoupon('DEMO25'),
+      ).toPromise();
+
+      expect(dispatched).toEqual([addCouponBegin(), addCouponFailure('uhoh')]);
+    });
+
+    it('should handle an unexpected error', async () => {
+      const errorResponse = 'oh snap!'; // We don't know how to handle this
+
+      const apiClientPost = jest.fn(() =>
+        new Promise((resolve, reject) => {
+          reject(errorResponse);
+        }));
+
+      configureApiService(
+        {
+          ECOMMERCE_BASE_URL: 'http://localhost',
+        },
+        {
+          post: apiClientPost,
+        },
+      );
+
+      const caughtErrors = [];
+      const dispatched = [];
+      try {
+        await runSaga(
+          {
+            dispatch: action => dispatched.push(action),
+            onError: error => caughtErrors.push(error),
+          },
+          handleAddCoupon,
+          addCoupon('DEMO25'),
+        ).toPromise();
+      } catch (e) {
+        expect(e).toEqual('oh snap!');
+      }
+
+      expect(caughtErrors).toEqual(['oh snap!']);
+    });
+  });
+
+  describe('handleRemoveCoupon', () => {
+    it('should handle success', async () => {
+      const responseData = {
+        data: {
+          order_total: 149,
+          calculated_discount: 12,
+          total_excl_discount: 161,
+        },
+      };
+
+      const apiClientDelete = jest.fn(() =>
+        new Promise((resolve) => {
+          resolve(responseData);
+        }));
+
+      configureApiService(
+        {
+          ECOMMERCE_BASE_URL: 'http://localhost',
+        },
+        {
+          delete: apiClientDelete,
+        },
+      );
+
+      const dispatched = [];
+      await runSaga(
+        { dispatch: action => dispatched.push(action) },
+        handleRemoveCoupon,
+        removeCoupon(12345),
+      ).toPromise();
+
+      expect(dispatched).toEqual([
+        removeCouponBegin(),
```

Comment:
Would it be better to compare the expected objects instead of calling the actions here? e.g. `{ type: REMOVE_COUPON.BEGIN }`

Reply:
I'm not sure; doing it that way feels like you're just copy-pasting the guts of the action in here, which feels a bit smelly to me.